### PR TITLE
Improve shims

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** feature

The following has been addressed in the PR:

* [X] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This PR will improve our shims, allowing them to be better integrated with TypeScript types and be able to be reduced dramatically when optimized with `dojo build` and static has flags.

It is currently a work in progress.  `Promise` and `Map` have been refactored.

### CODE REVIEW

File: src/Map.ts [hunk 1]
```diff
@@ -1,15 +1,53 @@
...
+	set(key: K, value: V): this;
+	readonly size: number;
+
+	/**
+	 * Returns an iterable of values in the map
+	 */
+	values(): IterableIterator<V>;
+}
+
+let ShimMap: MapConstructor;
```
Comment (agubler) [inline src/Map.ts]: Can you `let ShimMap: MapConstructor = global.Map;`

Then replace if no `es6-map` and then just `export default ShimMap;`?

Comment (kitsonk) [reply src/Map.ts]: Um... yes! DOH!

**Comment by kitsonk** [comment]
**breaking change** Several of our collection shims didn't include `undefined` returns, while the TypeScript libraries were updated.  This does mean in many situations we were potentially dealing with an undefined value when retrieving a value from a collection, so it should improve code, but it will cause regressions in several areas.

**Comment by kitsonk** [comment]
**breaking change** it is also necessary to either target `es2015` or beyond or include the `lib.es2015.promise` in the `lib`.  This will eliminate the problem of _duplicate Promise identifier_ errors though when targeting `es2015` and beyond in downstream projects, so it feels worth it.

File: src/Promise.ts [hunk 2]
```diff
@@ -19,39 +18,20 @@ export interface Executor<T> {
...
-	 *
-	 * @borrows Promise.all as Promise.all
-	 * @borrows Promise.race as Promise.race
-	 * @borrows Promise.reject as Promise.reject
-	 * @borrows Promise.resolve as Promise.resolve
-	 * @borrows Promise#catch as Promise#catch
-	 * @borrows Promise#then as Promise#then
-	 */
-	export class Promise<T> implements Thenable<T> {
+	global.Promise = ShimPromise = <any> class Promise<T> implements Thenable<T> {
```
Comment (agubler) [inline src/Promise.ts]: Why do you need to cast to `any` here?

**Pull request merged**